### PR TITLE
Update signatures for getAttributeValues member functions.

### PR DESCRIPTION
**Pull request by toregge**
@geirst : please review

### CODE REVIEW

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp [hunk 1]
```diff
@@ -271,12 +272,11 @@ DotProductExecutorBase<BaseType>::~DotProductExecutorBase() = default;
...
 void DotProductExecutorBase<BaseType>::execute(uint32_t docId) {
-    const AT *values(nullptr);
-    size_t count = getAttributeValues(docId, values);
-    size_t commonRange = std::min(count, _queryVector.size());
+    auto values = getAttributeValues(docId);
+    size_t commonRange = std::min(values.size(), _queryVector.size());
     static_assert(std::is_same<typename AT::ValueType, BaseType>::value);
     outputs().set_number(0, _multiplier.dotProduct(
-            &_queryVector[0], reinterpret_cast<const typename AT::ValueType *>(values), commonRange));
+            &_queryVector[0], reinterpret_cast<const typename AT::ValueType *>(&values[0]), commonRange));
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/dotproductfeature.cpp:279]: Consider adding _data()_ to ConstArrayRef to access the raw pointer.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp [hunk 2]
```diff
@@ -308,17 +310,16 @@ template <typename A>
...
-    const AT *allValues(NULL);
+    const AT* allValues(nullptr);
     size_t count = this->_attribute->getRawValues(docId, allValues);
-    values = &_scratch[0];
     size_t i(0);
     for (; (i < _queryIndexes.size()) && (_queryIndexes[i] < count); i++) {
         _scratch[i] = allValues[_queryIndexes[i]];
     }
-    return i;
+    return vespalib::ConstArrayRef(&_scratch[0], i);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/dotproductfeature.cpp:322]: Consider using std::vector::data() to access the raw pointer.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp [hunk 3]
```diff
@@ -332,16 +333,15 @@ template <typename A>
...
+DotProductByCopyExecutor<A>::getAttributeValues(uint32_t docId)
 {
     size_t count = this->_attribute->getAll(docId, &_copy[0], _copy.size());
     if (count > _copy.size()) {
         _copy.resize(count);
         count = this->_attribute->getAll(docId, &_copy[0], _copy.size());
     }
-    values = reinterpret_cast<const AT *>(&_copy[0]);
-    return count;
+    return vespalib::ConstArrayRef(reinterpret_cast<const AT *>(&_copy[0]), count);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/dotproductfeature.cpp:344]: Consider using std::vector::data() to access the raw pointer.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp [hunk 4]
```diff
@@ -367,8 +367,7 @@ SparseDotProductByCopyExecutor<A>::getAttributeValues(uint32_t docId, const AT *
     for (const IV & iv(this->_queryIndexes); (i < iv.size()) && (iv[i] < count); i++) {
         _copy[i] = _copy[iv[i]];
     }
-    values = reinterpret_cast<const AT *>(&_copy[0]);
-    return i;
+    return vespalib::ConstArrayRef(reinterpret_cast<const AT *>(&_copy[0]), i);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/dotproductfeature.cpp:370]: Consider using std::vector::data() to access the raw pointer.

**Review by geirst — APPROVED** [review]
👍

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp (hunk 1 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/features/dotproductfeature.cpp]: Done.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp (hunk 3 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/features/dotproductfeature.cpp]: Done.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp (hunk 4 above)
Comment (toregge) [reply searchlib/src/vespa/searchlib/features/dotproductfeature.cpp]: Done.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**